packages/common/src/hooks/useImageSize2.ts

```diff
@@ -0,0 +1,134 @@
+import { useState, useEffect, useCallback } from 'react'
+
+import { SquareSizes, Track } from '~/models'
+import { Maybe } from '~/utils/typeUtils'
+
+// Global image cache
+const IMAGE_CACHE = new Set<string>()
+
+/**
+ * Fetches an image from the given artwork object managing sizes and using fallback mirrors if necessary.
+ *  - If a larger image has already been fetched and is in the case, use it instead
+ *  - If a smaller image has already been fetched and is in the cache, use it while fetching the larger one
+ *  - If fetching a given image fails, substitute out mirrors and try again
+ * @param artwork - The artwork object to fetch from
+ * @param targetSize - The desired size of the image
+ * @param defaultImage - The fallback image to use if no image is found in the `artwork` object
+ * @returns The url of the image, or undefined if the image is not available
+ */
+export const useImageSize2 = ({
+  artwork,
+  targetSize,
+  defaultImage
+}: {
+  artwork: Maybe<Track['artwork']>
+  targetSize: SquareSizes
+  defaultImage: string
+}) => {
+  const [imageUrl, setImageUrl] = useState<string | null>(null)
+
+  const preloadImage = (url: string) => {
+    return new Promise((resolve, reject) => {
+      const img = new Image()
+      img.src = url
+      img.onload = () => resolve(url)
+      img.onerror = () => reject(url)
```

Comment:
It's a little curious that the reject case returns the URL instead of an error. In async/await flows, this would mean a `catch` block receives a string. I believe onerror might actually receive an error we could surface for debugging.

Reply:
Also wondering if the resolve case needs to return anything at all? Looks like we don't actually use the returned value below. And the code that calls this is passing the URL in as an argument and already knows the value.

Reply:
yeah, can not resolve or reject w/ any value. nice